### PR TITLE
DPSGD Mnist Example

### PR DESCRIPTION
**Pull request by cozek**
This PR adds an example where Flower is used to train a federeated learning system where clients are free to choose 
between non-private and private optimizers. Specifically, clients can choose to train Keras models using the standard SGD optimizer or __Differentially Private__ SGD (DPSGD) from [Tensorflow Privacy](https://github.com/tensorflow/privacy). For this task we use the MNIST dataset which is split artificially among clients. This causes the dataset to be i.i.d. The clients using DPSGD track the amount of privacy spent and display it at the end of the training. 

This example is adapted from https://github.com/tensorflow/privacy/blob/master/tutorials/mnist_dpsgd_tutorial_keras.py
In addition, i have used code in the examples/advanced_tensorflow and examples/simulation to build this example.

Due to lack of system resources, I am unable to run 10 clients at the same time and test like the other examples.
I have tested the scripts using 3 clients and they seem to work. Also, I am have used conda to manage the dependencies.

I would appreciate it if the reviewers would kindly test out the code.

### CODE REVIEW

File: examples/dp-sgd-mnist/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+description = "Federated training with Tensorflow Privacy"
+authors = [
+    "The Flower Authors <enquiries@flower.dev>",
+    "Kaushik Amar Das <kaushik.das@iiitg.ac.in>"
+]
+
+[tool.poetry.dependencies]
+python = "^3.8"
+tensorflow = "^2.5.0"
+tensorflow-privacy = {git = "https://github.com/tensorflow/privacy"}
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/pyproject.toml:19]: It's recommended to pin git dependencies to a specific commit (this avoids breaking changes from affecting our code example).

File: examples/dp-sgd-mnist/common.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    if noise_multiplier == 0.0:
+        return float('inf')
+    steps = epochs * num_train_examples // batch_size
+    orders = [1 + x / 10. for x in range(1, 100)] + list(range(12, 64))
+    sampling_probability = batch_size / num_train_examples
+    rdp = compute_rdp(q=sampling_probability,
+                      noise_multiplier=noise_multiplier,
+                      steps=steps,
+                      orders=orders)
+    # # Delta is set to approximate 1 / (number of training points).
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/common.py:35]: Duplicate `#`

File: examples/dp-sgd-mnist/common.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    return x[idx], y[idx]
+
+
+def partition(x: np.ndarray, y: np.ndarray, num_partitions: int) -> XYList:
+    """Split x and y into a number of partitions."""
+    return list(zip(np.array_split(x, num_partitions), np.array_split(y, num_partitions)))
+
+
+def preprocess(X: np.ndarray, y: np.ndarray) -> XY:
+    X = np.array(X, dtype=np.float32) / 255
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/common.py:71]: Docstring missing

File: examples/dp-sgd-mnist/client.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    if args.dpsgd and x_train.shape[0] % args.batch_size != 0:
+        drop_num = x_train.shape[0] % args.batch_size
+        x_train = x_train[:-drop_num]
+        y_train = y_train[:-drop_num]
+
+    # Start Flower client
+    client = MnistClient(model, x_train, y_train, x_test, y_test, args)
+    fl.client.start_numpy_client("[::]:8080", client=client)
+    if args.dpsgd:
+        print("Privacy Lost: ", PRIVACY_LOSS)
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/client.py:108]: Should the log message change here?
```suggestion
        print("Privacy Loss: ", PRIVACY_LOSS)
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
Thanks for the PR @cozek - it looks great!

I left comments on a few minor details, apart from that I'd say it's ready to be merged soon. Please also run the `black` formatter once (the CI doesn't check example formatting yet). If you've set the project up using Poetry, you can just run `poetry run ./dev/format.sh` from the project root.

File: examples/dp-sgd-mnist/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+poetry run python3 server.py &
+poetry run python3 client.py --partition 0 &
+poetry run python3 client.py --partition 1 --dpsgd True
+```
+
+It should be noted that when starting more than 2 clients, the total number of clients you intend to run and the data partition the client is expected to use must be specified. This is because the `num_clients` is used to split the dataset.
+For example, in case of 3 clients
+```shell
+poetry run python3 server.py --num-clients 3  &
+poetry run python3 server.py --num-clients 3 --partition 0 --dpsgd True &
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/README.md:73]:
```suggestion
poetry run python3 client.py --num-clients 3 --partition 0 --dpsgd True &
```

File: examples/dp-sgd-mnist/README.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+```shell
+poetry run python3 server.py &
+poetry run python3 client.py --partition 0 &
+poetry run python3 client.py --partition 1 --dpsgd True
+```
+
+It should be noted that when starting more than 2 clients, the total number of clients you intend to run and the data partition the client is expected to use must be specified. This is because the `num_clients` is used to split the dataset.
+For example, in case of 3 clients
+```shell
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/README.md:71]: Please add a blank line before this line

File: examples/dp-sgd-mnist/README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+Additional training parameters for the client and server can be referenced by passing `--help` to either script.
+
+Other things to note is that when all clients are running `dpsgd`, either train for more rounds or increase the local epochs to achieve optimal performance. You shall need to carefully tune the hyperparameters to your specific setup.
+```shell
+poetry run python3 server.py --num-clients 3  --num-rounds 20
+```
+
+```shell
+poetry run python3 client.py --partition 1 --local-epochs 4 --dpsgd True
```
Comment (danieljanes) [inline examples/dp-sgd-mnist/README.md:87]: Should we perhaps add `--num-clients 3` here as well?

**Review by danieljanes — CHANGES_REQUESTED** [review]
We ran the example with 10 clients and DPSGD enabled, they converge to ~89% test set accuracy.

There are a few more (minor) comments on the README, once they're resolved we can merge.

**Comment by cozek** [comment]
@danieljanes Thanks for testing it out. Maybe we can add a line at the end about the expected accuracy based on your testing with 10 clients.
```
Running this example with 10 clients and DPSGD enabled keeping all other parameters at their default,
should converge to ~89% test set accuracy.
```

**Review by danieljanes — DISMISSED** [review]
@cozek PR looks good now! Do you want to add the accuracy comment? We can merge either way

**Comment by cozek** [comment]
@danieljanes I have added the comment regarding test results. I am happy with it now and its good to merge.

**Review by danieljanes — APPROVED** [review]
